File: ios/Classes/ScatesdkFlutterPlugin.swift
import Flutter
import UIKit
import ScateSDK

public class ScatesdkFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var eventListeners: [String: NSObjectProtocol] = [:]
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "scatesdk_flutter", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "scatesdk_flutter_events", binaryMessenger: registrar.messenger())
        
        let instance = ScatesdkFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "Init":
            guard let args = call.arguments as? [String: Any],
                  let appId = args["appID"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing appID", details: nil))
                return
            }
            ScateCoreSDK.Init(appID: appId)
            result(nil)
        case "SetAdid":
            guard let args = call.arguments as? [String: Any],
                  let adid = args["adid"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing adid", details: nil))
                return
            }
            ScateCoreSDK.SetAdid(adid: adid)
            result(nil)
        case "Event":
            guard let args = call.arguments as? [String: Any],
                  let name = args["name"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing name", details: nil))
                return
            }
            ScateCoreSDK.Event(name: name)
            result(nil)
        case "EventWithValue":
            guard let args = call.arguments as? [String: Any],
                  let name = args["name"] as? String,
                  let value = args["value"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing name or value", details: nil))
                return
            }
            ScateCoreSDK.Event(name: name, customValue: value)
            result(nil)
        case "GetRemoteConfig":
            guard let args = call.arguments as? [String: Any],
                  let key = args["key"] as? String,
                  let defaultValue = args["defaultValue"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing key or defaultValue", details: nil))
                return
            }
            let value = ScateCoreSDK.GetRemoteConfig(key: key, defaultValue: defaultValue)
            result(value)
        case "AddListener":
            guard let args = call.arguments as? [String: Any],
                  let name = args["name"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing listener name", details: nil))
                return
            }
            addListener(name)
            result(nil)
        case "RemoveListener":
            guard let args = call.arguments as? [String: Any],
                  let name = args["name"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing listener name", details: nil))
                return
            }
            removeListener(name)
            result(nil)
        case "OnboardingStart":
            ScateCoreSDK.OnboardingStart()
            result(nil)
            
        case "OnboardingStep":
            guard let args = call.arguments as? [String: Any],
                  let step = args["step"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing step", details: nil))
                return
            }
            ScateCoreSDK.OnboardingStep(step: step)
            result(nil)
            
        case "OnboardingFinish":
            ScateCoreSDK.OnboardingFinish()
            result(nil)
            
        case "LoginSuccess":
            guard let args = call.arguments as? [String: Any],
                  let source = args["source"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing source", details: nil))
                return
            }
            ScateCoreSDK.LoginSuccess(source: source)
            result(nil)
            
        case "InterstitialAdShown":
            ScateCoreSDK.InterstitialAdShown()
            result(nil)
            
        case "InterstitialAdClosed":
            ScateCoreSDK.InterstitialAdClosed()
            result(nil)
            
        case "RewardedAdShown":
            ScateCoreSDK.RewardedAdShown()
            result(nil)
            
        case "RewardedAdClosed":
            ScateCoreSDK.RewardedAdClosed()
            result(nil)
            
        case "RewardedAdClaimed":
            ScateCoreSDK.RewardedAdClaimed()
            result(nil)
            
        case "BannerAdShown":
            ScateCoreSDK.BannerAdShown()
            result(nil)
            
        case "NotificationPermissionGranted":
            ScateCoreSDK.NotificationPermissionGranted()
            result(nil)
            
        case "NotificationPermissionDenied":
            ScateCoreSDK.NotificationPermissionDenied()
            result(nil)
            
        case "LocationPermissionGranted":
            ScateCoreSDK.LocationPermissionGranted()
            result(nil)
            
        case "LocationPermissionDenied":
            ScateCoreSDK.LocationPermissionDenied()
            result(nil)
            
        case "ATTPromptShown":
            ScateCoreSDK.ATTPromptShown()
            result(nil)
            
        case "ATTPermissionGranted":
            ScateCoreSDK.ATTPermissionGranted()
            result(nil)
            
        case "ATTPermissionDenied":
            ScateCoreSDK.ATTPermissionDenied()
            result(nil)
            
        case "PaywallShown":
            guard let args = call.arguments as? [String: Any],
                  let paywall = args["paywall"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing paywall", details: nil))
                return
            }
            ScateCoreSDK.PaywallShown(paywall: paywall)
            result(nil)
            
        case "PaywallClosed":
            guard let args = call.arguments as? [String: Any],
                  let paywall = args["paywall"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing paywall", details: nil))
                return
            }
            ScateCoreSDK.PaywallClosed(paywall: paywall)
            result(nil)
            
        case "PaywallAttempted":
            guard let args = call.arguments as? [String: Any],
                  let paywall = args["paywall"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing paywall", details: nil))
                return
            }
            ScateCoreSDK.PaywallAttempted(paywall: paywall)
            result(nil)
            
        case "PaywallPurchased":
            guard let args = call.arguments as? [String: Any],
                  let paywall = args["paywall"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing paywall", details: nil))
                return
            }
            ScateCoreSDK.PaywallPurchased(paywall: paywall)
            result(nil)
            
        case "PaywallCancelled":
            guard let args = call.arguments as? [String: Any],
                  let paywall = args["paywall"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing paywall", details: nil))
                return
            }
            ScateCoreSDK.PaywallCancelled(paywall: paywall)
            result(nil)
            
        case "TabClicked":
            guard let args = call.arguments as? [String: Any],
                  let tab = args["tab"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing tab", details: nil))
                return
            }
            ScateCoreSDK.TabClicked(tab: tab)
            result(nil)
            
        case "FeatureClicked":
            guard let args = call.arguments as? [String: Any],
                  let feature = args["feature"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing feature", details: nil))
                return
            }
            ScateCoreSDK.FeatureClicked(feature: feature)
            result(nil)
            
        case "DailyStreakShown":
            ScateCoreSDK.DailyStreakShown()
            result(nil)
            
        case "DailyStreakClaimed":
            ScateCoreSDK.DailyStreakClaimed()
            result(nil)
            
        case "DailyStreakClosed":
            ScateCoreSDK.DailyStreakClosed()
            result(nil)   

        case "ShowPaywall":
            guard let args = call.arguments as? [String: Any],
                  let jsonString = args["jsonString"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing paywall", details: nil))
                return
            }
            ScateCoreSDK.ShowPaywall(jsonString: jsonString)
            result(nil) 
        case "ShowOnboarding":
            guard let args = call.arguments as? [String: Any],
                  let jsonString = args["jsonString"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing paywall", details: nil))
                return
            }
            ScateCoreSDK.ShowOnboarding(jsonString: jsonString)
            result(nil)
        case "ClosePaywall":
            ScateCoreSDK.ClosePaywall()
            result(nil)   
        case "CloseOnboarding":
            ScateCoreSDK.CloseOnboarding()
            result(nil)  
        case "ClosePaidProductLoadingScreen":
            ScateCoreSDK.ClosePaidProductLoadingScreen()
            result(nil)  
        case "ShowEventList":
            ScateCoreSDK.ShowEventList()
            result(nil) 
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    private func addListener(_ name: String) {
        let observer = NotificationCenter.default.addObserver(forName: Notification.Name(name), object: nil, queue: nil) { [weak self] notification in
            self?.handleListener(notification: notification)
        }
        eventListeners[name] = observer
    }
    
    private func removeListener(_ name: String) {
        if let observer = eventListeners[name] {
            NotificationCenter.default.removeObserver(observer)
            eventListeners[name] = nil
        }
    }
    
    @objc func handleListener(notification: Notification) {
        var event: [String: Any] = [:]
   
        do{
            let eventName = notification.name.rawValue
            event["event"] = eventName
            event["data"] = notification.userInfo
            
            do {
                let eventData = try JSONSerialization.data(withJSONObject: event, options: [])
                if let jsonString = String(data: eventData, encoding: .utf8) {
                    DispatchQueue.main.async {
                        // Ensure eventSink operation is on main thread
                        self.eventSink?(jsonString)
                    }
                }
            } catch {
                print("Error converting event data to JSON: \(error.localizedDescription)")
            }
        }
    }
}

File: android/src/main/kotlin/com/scate/scatesdk_flutter/ScatesdkFlutterPlugin.kt
package com.scate.scatesdk_flutter

import android.content.Context
import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.EventChannel.StreamHandler
import com.scate.scatesdk.ScateCoreSDK
import com.scate.scatesdk.models.RemoteConfigListener;
import org.json.JSONObject
import android.os.Handler
import android.os.Looper

/** ScatesdkFlutterPlugin */
class ScatesdkFlutterPlugin: FlutterPlugin, MethodCallHandler, StreamHandler {

  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private lateinit var context: Context
  private var eventSink: EventSink? = null

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "scatesdk_flutter")
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "scatesdk_flutter_events")
    methodChannel.setMethodCallHandler(this)
    eventChannel.setStreamHandler(this)
    context = flutterPluginBinding.applicationContext // Obtain application context here

    val remoteConfigListener = PluginRemoteConfigListener()
    ScateCoreSDK.addRemoteConfigListener(remoteConfigListener)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "Init" -> {
                val appID: String? = call.argument("appID")
                ScateCoreSDK.init(appID, context)
                result.success(null)
            }
            "SetAdid" -> {
                val adid: String? = call.argument("adid")
                ScateCoreSDK.SetAdid(adid)
                result.success(null)
            }
            "Event" -> {
                val name: String? = call.argument("name")
                ScateCoreSDK.event(name)
                result.success(null)
            }
            "EventWithValue" -> {
                val name: String? = call.argument("name")
                val value: String? = call.argument("value")
                ScateCoreSDK.event(name, value)
                result.success(null)
            }
            "GetRemoteConfig" -> {
                val key: String? = call.argument("key")
                val defaultValue: String? = call.argument("defaultValue")
                val value = ScateCoreSDK.getRemoteConfig(key, defaultValue)
                result.success(value)
            }
            "AddListener" -> {
                result.success(null)
            }
            "RemoveListener" -> {
                result.success(null)
            }
                "OnboardingStart" -> {
            ScateCoreSDK.OnboardingStart()
            result.success(null)
            }
            "OnboardingStep" -> {
                val step = call.argument<String>("step") ?: run {
                    result.error("INVALID_ARGUMENT", "Missing step", null)
                    return
                }
                ScateCoreSDK.OnboardingStep(step)
                result.success(null)
            }
            "OnboardingFinish" -> {
                ScateCoreSDK.OnboardingFinish()
                result.success(null)
            }
            "LoginSuccess" -> {
                val source = call.argument<String>("source") ?: run {
                    result.error("INVALID_ARGUMENT", "Missing source", null)
                    return
                }
                ScateCoreSDK.LoginSuccess(source)
                result.success(null)
            }
            "InterstitialAdShown" -> {
                ScateCoreSDK.InterstitialAdShown()
                result.success(null)
            }
            "InterstitialAdClosed" -> {
                ScateCoreSDK.InterstitialAdClosed()
                result.success(null)
            }
            "RewardedAdShown" -> {
                ScateCoreSDK.RewardedAdShown()
                result.success(null)
            }
            "RewardedAdClosed" -> {
                ScateCoreSDK.RewardedAdClosed()
                result.success(null)
            }
            "RewardedAdClaimed" -> {
                ScateCoreSDK.RewardedAdClaimed()
                result.success(null)
            }
            "BannerAdShown" -> {
                ScateCoreSDK.BannerAdShown()
                result.success(null)
            }
            "NotificationPermissionGranted" -> {
                ScateCoreSDK.NotificationPermissionGranted()
                result.success(null)
            }
            "NotificationPermissionDenied" -> {
                ScateCoreSDK.NotificationPermissionDenied()
                result.success(null)
            }
            "LocationPermissionGranted" -> {
                ScateCoreSDK.LocationPermissionGranted()
                result.success(null)
            }
            "LocationPermissionDenied" -> {
                ScateCoreSDK.LocationPermissionDenied()
                result.success(null)
            }
            "ATTPromptShown" -> {
                ScateCoreSDK.ATTPromptShown()
                result.success(null)
            }
            "ATTPermissionGranted" -> {
                ScateCoreSDK.ATTPermissionGranted()
                result.success(null)
            }
            "ATTPermissionDenied" -> {
                ScateCoreSDK.ATTPermissionDenied()
                result.success(null)
            }
            "PaywallShown" -> {
                val paywall = call.argument<String>("paywall") ?: run {
                    result.error("INVALID_ARGUMENT", "Missing paywall", null)
                    return
                }
                ScateCoreSDK.PaywallShown(paywall)
                result.success(null)
            }
            "PaywallClosed" -> {
                val paywall = call.argument<String>("paywall") ?: run {
                    result.error("INVALID_ARGUMENT", "Missing paywall", null)
                    return
                }
                ScateCoreSDK.PaywallClosed(paywall)
                result.success(null)
            }
            "PaywallAttempted" -> {
                val paywall = call.argument<String>("paywall") ?: run {
                    result.error("INVALID_ARGUMENT", "Missing paywall", null)
                    return
                }
                ScateCoreSDK.PaywallAttempted(paywall)
                result.success(null)
            }
            "PaywallPurchased" -> {
                val paywall = call.argument<String>("paywall") ?: run {
                    result.error("INVALID_ARGUMENT", "Missing paywall", null)
                    return
                }
                ScateCoreSDK.PaywallPurchased(paywall)
                result.success(null)
            }
            "PaywallCancelled" -> {
                val paywall = call.argument<String>("paywall") ?: run {
                    result.error("INVALID_ARGUMENT", "Missing paywall", null)
                    return
                }
                ScateCoreSDK.PaywallCancelled(paywall)
                result.success(null)
            }
            "TabClicked" -> {
                val tab = call.argument<String>("tab") ?: run {
                    result.error("INVALID_ARGUMENT", "Missing tab", null)
                    return
                }
                ScateCoreSDK.TabClicked(tab)
                result.success(null)
            }
            "FeatureClicked" -> {
                val feature = call.argument<String>("feature") ?: run {
                    result.error("INVALID_ARGUMENT", "Missing feature", null)
                    return
                }
                ScateCoreSDK.FeatureClicked(feature)
                result.success(null)
            }
            "DailyStreakShown" -> {
                ScateCoreSDK.DailyStreakShown()
                result.success(null)
            }
            "DailyStreakClaimed" -> {
                ScateCoreSDK.DailyStreakClaimed()
                result.success(null)
            }
            "DailyStreakClosed" -> {
                ScateCoreSDK.DailyStreakClosed()
                result.success(null)
            }
             else -> result.notImplemented()
        }
    }

  // StreamHandler methods
  override fun onListen(arguments: Any?, events: EventSink?) {
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }

  inner class PluginRemoteConfigListener : RemoteConfigListener {
    override fun onRemoteConfigInitialized(success: Boolean) {
        println("PluginRemoteConfigListener")
        val event = mutableMapOf<String, Any>()

        // Event name
        val eventName = "Scate_RemoteConfigsReady"
        event["event"] = eventName

        // Event data (success status)
        val data = mutableMapOf<String, Any>()
        data["remoteConfigFetched"] = success
        event["data"] = data

        try {
            // Convert the event map to a JSON string
            val eventJson = JSONObject(event as Map<*, *>).toString()

            // Ensure the event is sent on the main thread
            Handler(Looper.getMainLooper()).post {
                eventSink?.success(eventJson)
            }
            
        } catch (e: Exception) {
            println("Error converting event data to JSON: ${e.localizedMessage}")
        }
    }
}
}
